tfhe/src/integer/server_key/radix_parallel/add.rs

```diff
@@ -386,188 +398,956 @@ impl ServerKey {
+            let num_blocks = lhs.len();
+
+            // We did the first block before, the last block is done after this if,
+            // so we need 3 blocks at least to enter this
+            if num_blocks >= 3 {
+                for (lhs_b, rhs_b) in lhs[1..num_blocks - 1]
+                    .iter_mut()
+                    .zip(rhs[1..num_blocks - 1].iter())
+                {
+                    self.key.unchecked_add_assign(lhs_b, rhs_b);
+                    self.key.unchecked_add_assign(lhs_b, &carry);
+
+                    carry.clone_from(lhs_b);
+                    rayon::join(
+                        || self.key.message_extract_assign(lhs_b),
+                        || self.key.carry_extract_assign(&mut carry),
+                    );
+                }
+            }
+
+            if num_blocks >= 2 {
+                // Handle the last block
+                self.key
+                    .unchecked_add_assign(&mut lhs[num_blocks - 1], &rhs[num_blocks - 1]);
+                self.key
+                    .unchecked_add_assign(&mut lhs[num_blocks - 1], &carry);
+            }
+
+            if let Some(block) = overflow_flag.as_mut() {
+                if num_blocks == 1 && input_carry.is_some() {
+                    self.key
+                        .unchecked_add_assign(block, input_carry.map(|b| &b.0).unwrap());
+                } else {
+                    self.key.unchecked_add_assign(block, &carry);
+                }
+            }
+
+            // To be able to use carry_extract_assign in it
+            carry.clone_from(&lhs[num_blocks - 1]);
+
+            // Note that here when num_blocks == 1 && requested_flag != Overflow nothing
+            // will actually be spawned.
+            rayon::scope(|s| {
+                if num_blocks >= 2 {
+                    // These would already have been done when the first block was processed
+                    s.spawn(|_| {
+                        self.key.message_extract_assign(&mut lhs[num_blocks - 1]);
+                    });
+
+                    s.spawn(|_| {
+                        self.key.carry_extract_assign(&mut carry);
+                    });
+                }
+
+                if requested_flag == OutputFlag::Overflow {
+                    s.spawn(|_| {
+                        let overflow_flag_block = overflow_flag.as_mut().unwrap();
+                        //let shifted_carry = self.key.unchecked_scalar_mul(&carry, 2);
+                        // Computing the overflow flag requires and extra step for the first block
+                        let overflow_flag_lut = self.key.generate_lookup_table(|block| {
+                            let input_carry = block & 1;
+                            if input_carry == 1 {
+                                (block >> 3) & 1
+                            } else {
+                                (block >> 2) & 1
+                            }
```

Comment:
Its the result from the lines 

```
        (overflows_with_given_input_carry(1) << 3)
                                | (overflows_with_given_input_carry(0) << 2)
```

at bit index 3 we stored a value (0 or 1) that tells whether overflow occured if the input carry of the block were to be 1,
at bit index 2, we stored the same kind of value but for when the carry is 0


So this lust receives a block where bit 2 & 3 have the values described above, and in its bit index 0, it has the input carry of the block, so it then can return if overflow happens

Reply:
ok a small comment somewhere for the encoding is of interest I think

do we need to shift by 2 and 3 though ? shifting by 1 and 0 (so no shift) would be enough or I guess I'm missing something related to the carry prop ?

Reply:
I think we can shift by 1 && 2 (0 is where the input carry is stored)
although for the scalar versions of these we might need to shift by 2 && 3

Reply:
And I'd want to try something that would allow the sequential carry propagation of groups to be done 4 by 4 and it would require to shift by 2 && 3 this value

its not as critical as it woudn't bring improvements to 32 & 64 bits, bit for sizes in between it should,

Reply:
ok no worries, keep the current encoding then